### PR TITLE
xplr: fix cfg.plugins implementation for more than one plugin

### PR DESCRIPTION
**Pull request by ChanceHarrison**
### Description

Would close https://github.com/nix-community/home-manager/issues/4520. Please see that issue (and the upstream discussion linked within) for more information.

Change the implementation of the `programs.xplr.plugins` option to correctly set the `package.path` in the written Lua config file.

Specifically, each component path of the `package.path` must have appropriate wildcard/template characters (what Lua calls "interrogation marks" or `?` characters).

Furthermore, due to the need (in the current implementation) to use `programs.xplr.extraConfig` to call `require(<plugin-name>).setup`, this implementation ensures that plugins in the Nix store are inside a predictably-named directory.

### Checklist

- [ ] Change is backwards compatible.

Unfortunately not. The `cfg.plugins` option currently takes a list for while it is proposed that it takes an attribute set.

While a change could be made that maintains complete backwards compatibility, one could argue that the usability improvements that accompany the change (of an already broken option) outweigh the cost of breaking back-compat.

- [x] Code formatted with ~~`./format`~~ `nixfmt`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted appropriately

#### Maintainer CC

@NotAShelf

### CODE REVIEW

File: modules/programs/xplr.nix
Comment (sayanarijit) [inline modules/programs/xplr.nix:66]: I'd use fetchGit here, for people not using git.

File: modules/programs/xplr.nix
Comment (sayanarijit) [inline modules/programs/xplr.nix:22]:
```suggestion
          mkdir './${name}'
          mv $out/* './${name}/'
          mv './${name}' $out/
```

**Review by sayanarijit — CHANGES_REQUESTED** [review]
Minor suggestions. Other than that, pr looks good. cc @NotAShelf for further review.

**Comment by NotAShelf** [comment]
> Minor suggestions. Other than that, pr looks good. cc @NotAShelf for further review.

I'm in favor of approving this PR as is (with our without the changes you've mentioned) but I *really* would like some second opinion on the plugin wrapper. I'll try communicating on matrix about this.

File: modules/programs/xplr.nix
Comment (NotAShelf) [reply modules/programs/xplr.nix:22]: is mkdir even necessary here?

Comment (sayanarijit) [reply modules/programs/xplr.nix:22]: We need to make a dir named "${name}" so that we can move the contents in there.

**Review by NotAShelf — APPROVED** [review]

File: modules/programs/xplr.nix
Comment (eclairevoyant) [inline modules/programs/xplr.nix:28]: This whole wrapping thing seems strange, what happens if you wrap a plugin in a store path that doesn't have init.lua? I assume it won't match because it'd be `${p}/?/?.lua` not just `${p}/?.lua`

Alternatively, we only allow plugins with `init.lua` (and we can get rid of all this wrapping and wildcard fluff).

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: modules/programs/xplr.nix
Comment (sayanarijit) [reply modules/programs/xplr.nix:28]: With only ${p}/?.lua, we'd need to import the plugin as `name.init`, assuming the entrypoint is `init.lua`. It's a lua convention to add init.lua in package path to make the lib easy to import.

Comment (eclairevoyant) [reply modules/programs/xplr.nix:28]: Then I think my original comment applies?

> what happens if you wrap a plugin in a store path that doesn't have init.lua? I assume it won't match because it'd have to be `${p}/?/?.lua` not just `${p}/?.lua

Comment (sayanarijit) [reply modules/programs/xplr.nix:28]: The wrapping does this:
/nix/store/hash/init.lua ->
/nix/store/hash/name/init.lua

So we can add add `package.path = "/nix/store/hash/?/.init.lua;/nix/store/hash/*.lua"` and `require("name")`.

To understand the issue better, let's reproduce it.

1. Create /fakenixstore
2. Git clone sayanarijit/trash-cli.xplr as /fakenixstore/hashsha123 such that sayanarijit/trash-cli.xplr/init.lua is now /fakenixstore/hashsha123/init.lua
3. Get into lua repl (in any dir) and modify package.path in a way so that you can run `require("trash-cli").setup()`.

Here you'll see that the only way to make it work is to:
1. move or symlink `/fakenixstore/hashsha123/*` as `/fakenixstore/hashsha123/trash-cli/*`.
2. In lua repl, run `package.path = "/fakenixstore/hashsha123/?/.init.lua;/nix/fakenixstorestore/hashsha123/*.lua"`

File: modules/programs/xplr.nix
Comment (rycee) [reply modules/programs/xplr.nix:66]: Agreed, although I would suggest `fetchFromGitHub`. If you search for that function you'll find already existing examples.

File: modules/programs/xplr.nix
Comment (rycee) [inline modules/programs/xplr.nix:59]: Perhaps you mean
```suggestion
        An attribute set of plugin paths to be added to the [package.path](https://www.lua.org/manual/5.4/manual.html#pdf-package.path) of the {file}`~/config/xplr/init.lua` configuration file.
```
?

Also, I would suggest wrapping the line at roughly 80 char width.

Comment (ChanceHarrison) [reply modules/programs/xplr.nix:59]: Thank you. Resolved by latest force push.

File: modules/programs/xplr.nix
Comment (ChanceHarrison) [reply modules/programs/xplr.nix]: Thank you. Resolved by latest force push.

File: modules/programs/xplr.nix
Comment (ChanceHarrison) [reply modules/programs/xplr.nix]: Thank you. Resolved by latest force push.

File: modules/programs/xplr.nix
Comment (ChanceHarrison) [reply modules/programs/xplr.nix]: Thank you. Resolved by latest force push.

File: modules/programs/xplr.nix
Comment (ChanceHarrison) [reply modules/programs/xplr.nix:28]: @eclairevoyant Based on the most recent message, is it clear why the wrapping method is being used? Do you see any better way? If I or someone can provide more clarification, just say the word. Otherwise, feel free to resolve the conversation.

**Comment by ChanceHarrison** [comment]
> @NotAShelf  I'm in favor of approving this PR as is (with our without the changes you've mentioned) but I _really_ would like some second opinion on the plugin wrapper. I'll try communicating on matrix about this.

Thanks for the review. Did anything come out of your communications on Matrix regarding the plugin wrapping?

---

To all: Aside from the unresolved conversation regarding the plugin wrapping, I am moving this PR out of draft status based on the feedback received. If there is anything else you would like addressed, just let me know.

**Comment by NotAShelf** [comment]
I haven't received any responses on matrix but rycee has already given a review - I also asked a few other people to provide their own.

I'll go ahead and approve this.

**Review by NotAShelf — APPROVED** [review]

**Review by fufexan — APPROVED** [review]

**Review by sayanarijit — APPROVED** [review]

**Comment by sayanarijit** [comment]
Hi @rycee I think this is good to go, if no-one has any concern.

**Comment by rycee** [comment]
Thanks! Merged to master now. If there is any residual concerns about the wrapper, perhaps those could be resolved in a future PR.

**Pull request merged**
